File: Coding-interview/cracking-coding-interview/chapter-1/golang/1.5_one_away.go
package main

import "fmt"

const INSERT_TYPE = "insert"
const REMOVE_TYPE = "remove"

func oneEditReplace(s1 string, s2 string) bool {
	foundDiff := false
	for index := 0; index < len(s1); index++ {
		if s1[index] != s2[index] {
			if foundDiff {
				return false
			}
			foundDiff = true
		}
	}
	return true
}

func oneEditInsertAndRemove(s1 string, s2 string, operationType string) bool {
	firstStringIndex := 0
	secondStringIndex := 0
	for firstStringIndex < len(s1) && secondStringIndex < len(s2) {
		if s1[firstStringIndex] != s2[secondStringIndex] {
			if firstStringIndex != secondStringIndex {
				return false
			}
			if operationType == INSERT_TYPE {
				firstStringIndex++
			} else if operationType == REMOVE_TYPE {
				secondStringIndex++
			}
		} else {
			firstStringIndex++
			secondStringIndex++
		}
	}
	return true
}

func oneEditAway(s1 string, s2 string) bool {
	if len(s1) == len(s2) {
		return oneEditReplace(s1, s2)
	} else if len(s1)+1 == len(s2) {
		return oneEditInsertAndRemove(s1, s2, REMOVE_TYPE)
	} else if len(s1)-1 == len(s2) {
		return oneEditInsertAndRemove(s1, s2, INSERT_TYPE)
	}
	return false
}

func main() {
	result := oneEditAway("pale", "ple")
	fmt.Println(result)
}

File: Coding-interview/cracking-coding-interview/chapter-1/c++/1.5_one_away.cpp
# include <iostream>
# include <cstdio>
# include <cstring>

using namespace std;

/*
  ale -> aple
  apple -> aple
*/
	
bool onEditInsert(string s1, string s2) {
	int firstStringIndex = 0, secondStringIndex = 0;
	while(firstStringIndex < s1.size() && secondStringIndex < s2.size()) {
		if (s1[firstStringIndex] != s2[secondStringIndex]) {
			if(firstStringIndex != secondStringIndex) {
				return false;
			}
			secondStringIndex++;
		}
		else {
			firstStringIndex++;
			secondStringIndex++;
		}
	}  
	return true;
}

bool onEditReplace(string s1, string s2) {
	bool foundDiff = false;
	for (int index = 0 ; index < s1.size(); index++) {
		if (s1[index] != s2[index]) {
			if (foundDiff) {
				return false;
			}
			foundDiff = true;
		}
	}
	return true;
}

bool onEditAway(string s1, string s2) {
	if (s1.size() == s2.size()) {
		return onEditReplace(s1, s2);
	}
	else if (s1.size() + 1 == s2.size()) {
		return onEditInsert(s1, s2);
	}
	else if (s1.size() - 1 == s2.size()) {
		return onEditInsert(s1, s2);
	}
	return false;
}



int main() {
	freopen("in.txt", "r", stdin);
	string s1, s2;
	cin >>s1 >>s2;
	bool result = onEditAway(s1, s2);
	printf("%d\n", result);
	return 0;
}